src/alarms/alarm-handling.md

```diff
@@ -0,0 +1,102 @@
+| message-string    | Human readable short information about alarm reason (e.g. "Temperature sensor does not respond") |
+| severity-string   | Could be "CRITICAL", "MAJOR", "MINOR" or "WARNING" |
+| status-string     | Could be "ACTIVE" or "CLEARED"<br/>
+| time-string       | Timestamp indicating when the alarm (or the alarm-update) had occured (in [ISO 8601 format](http://www.w3.org/TR/NOTE-datetime#)) |
+
+# Data flow
+
+Figure below illustrates the data flow from Customer Application broker/thin-edge up to cloud. It shows especially the use of MQTT QOS=1, MQTT Retain and Mosquitto Persistence to achieve reliability alarm state-change transfer.
+
+![Sequence Diagram Update SW-list](images/alarm_dataflow.svg)
+
+**To be decided:** Interface from Mapper to Cloud to be defined. Two options are possible:
+1) **JSONviaMQTT:** To have reliability use QOS=1. Open issue: Not yet completely confirmed that C8Y sends PUBACK when msg was processed, instead just on arrival.
+2) **HTTP REST:** HTTP response indicates if message was processed sucessfully. Anyway HTTP REST would increase complexity of implementation and increase data traffic.
+
+## Requirements
+|   |
+|---------------|
+| Thin-edge shall transfer most recent state of an alarm to the cloud. |
+| Other local components (as local processes or connected clients) can consume alarm-states from thin-edge by subscribing local broker topics. A local consumer can be made to process full alarm-state history by using "Clean Start Flag=0", a constant "client Id" and "QOS>0". |
+| The same alarm-state message shall be transfered just once to the cloud.<br/>An alarm-state message is treated as different when at least one field (see section 'Information Set per Alarm' above) in the message differs to the message before.<br/>To avoid duplicates some mapper-specific topic shall be used (retained) to store and identify messages that were already transfered to the cloud (e.g. `mapper/c8y/ack/alarms/<alarm>`). |
+
+
+# Public MQTT-based alarm interface
+
+Similar to thin-edge'S measurement interface, alarm interface is based on MQTT topics.
+
+## Topic structure and payload
+
+Proposal to have "severity" and "type-string" as topics:
+
+**thin-edge JSON (alarm) format:**
+```
+Topic:   tedge/alarms/<severity-string>/<type-string>
+Payload: {
+  "message":  <message-string>,
+  "status":   <status-string>,  
+  "time":     <time-string>
+}
+```
+
+NOTE: The sub-topic `childs` can not be used as type-string, since it is reseved for child-device addressing (see below).
+
+**Adressing Child-Devices:**
+To address an alarm to a child-device the sub-topic `childs` followed by the child's device id has to be used as below:
+```
+Child-Device Topics:   tedge/alarms/childs/<child-device id>/<severity-string>/<type-string>
+```
+
+Benefit to have "severity" and "type-string" as topics:

```

Comment:
I think the criteria to put or not an attribute of the alarm inside the topic name is not about filtering alarms but about the identities of these alarms. Since we are using retain flags we must have a one-one relationship between a topic name and an alarm.

The right question is what identify *uniquely* an alarm as an *entity*:
* The type of the alarm and its location (the device, a child device or a source): definitely.
* But the severity of the alarm? Maybe. It depends if you consider or not "critical/high-temperature" as the *same* alarm or not as "minor/high-temperature".
* The timestamp is definitely not part of the identifier of the alarm: the *same* alarm can been raised several time. Similarly, one can consider that the *same* alarm can be raised at different severity level over time. In such a case, I would not put the severity level in the topic name.

So the main question to answer before addressing topic naming is :
>  Are "critical/high-temperature" and "minor/high-temperature"
>  * two different alarms ?
>  * or two different events for the same alarm?

Reply:
In my definition of an alarm it has a fixed severity. So my answer to your question: 
-> "critical/high-temperature" and "minor/high-temperature" are two different alarms.

By the way: Having a fixed severity for each alarm is the way I know from PLCs. E.g. in CoDeSys you can define an Alarm with just a fixed severity. Also Siemens SIMATIC Panels treat it like that.

Reply:
In that case, a topic name that includes the severity makes sense. This is even the way to go.